index/slashing/slashing.go

```diff
@@ -41,167 +48,186 @@ type SlashingIndex struct {
+		history := make([]uint64, len(v.Epochs))
+		copy(history, v.Epochs)
+		ii.Miners[addr] = Slashes{
+			Epochs: history,
 		}
 	}
 	return ii
 }
+
+// Listen returns a a signaler channel which signals that index information
+// has been updated.
 func (s *SlashingIndex) Listen() <-chan struct{} {
 	return s.signaler.Listen()
 }
 
+// Unregister frees a channel from the signaler hub
 func (s *SlashingIndex) Unregister(c chan struct{}) {
 	s.signaler.Unregister(c)
 }
 
-// ToDo: load from datastore
+// Close closes the SlashindIndex
+func (s *SlashingIndex) Close() error {
+	log.Info("Closing")
+	s.clsLock.Lock()
+	defer s.clsLock.Unlock()
+	if s.closed {
+		return nil
+	}
+	s.cancel()
+	<-s.finished
+	s.closed = true
+	return nil
+}
+
+// start is a long running job that keeps the index up to date with chain updates
 func (s *SlashingIndex) start() {
 	defer close(s.finished)
-	n, err := s.c.ChainNotify(s.ctx)
+	n, err := s.api.ChainNotify(s.ctx)
 	if err != nil {
 		log.Fatalf("error when getting notify channel from lotus: %s", err)
 	}
 	for {
 		select {
 		case <-s.ctx.Done():
-			log.Info("graceful shutdown of background storage reputation builder")
+			log.Info("graceful shutdown of background slashing updater")
 			return
-		case _, ok := <-n:
+		case hcs, ok := <-n:
 			if !ok {
-				log.Fatalf("lotus notify channel closed: %s", err)
+				log.Error("lotus notify channel closed")
+				return
 			}
 			log.Info("updating slashing index...")
-			if err := s.update(); err != nil {
-				log.Errorf("error when updating storage reputation: %s", err)
+			currHeadTsk := types.NewTipSetKey(hcs[len(hcs)-1].Val.Cids...)
+			if err := s.updateIndex(currHeadTsk); err != nil {
+				log.Errorf("error when updating slashing history: %s", err)
 				continue
 			}
 			log.Info("slashing index updated")
 		}
 	}
 }
 
-func (s *SlashingIndex) update() error {
-	base, err := s.lastSyncedTipset()
+// updateIndex updates current index with a new discovered chain head.
+func (s *SlashingIndex) updateIndex(new types.TipSetKey) error {
+	var index Index
+	ts, err := s.store.LoadAndPrune(s.ctx, new, &index)
 	if err != nil {
 		return err
 	}
-	path, err := chainsync.GetPathToHead(s.ctx, s.c, base)
+	if index.Miners == nil {
+		index.Miners = make(map[string]Slashes)
+	}
+	_, path, err := chainsync.ResolveBase(s.ctx, s.api, ts, new)
 	if err != nil {
 		return err
 	}
-	if err = s.updatePath(path); err != nil {
-		return fmt.Errorf("error update history: %s", err)
-	}
-	if err := s.save(); err != nil {
-		return err
+	mctx := context.Background()
+	start := time.Now()
+	for i := 0; i < len(path); i += batchSize {
+		j := i + batchSize
+		if j > len(path) {
+			j = len(path)
+		}
+		if err := updateFromPath(s.ctx, s.api, &index, path[i:j]); err != nil {
+			return err
+		}
+		if err := s.store.Save(s.ctx, types.NewTipSetKey(path[j-1].Cids...), index); err != nil {
+			return err
+		}
+		stats.Record(mctx, mRefreshProgress.M(float64(i)/float64(len(path))))
 	}
```

Comment:
Here's the most powerful leverage of `chainstore` and `chainsync`.
`chainstore` is use similar as said before: get last same-fork saved state from newly discovered Tipset.
Having this saved state as a starting point, we can use `ResolveBase`, which will return us a path from this base tipset state to `new`. (This is a simplified case since we know that saved state is from the same fork, so the path returned doesn't have any `revert` actions, only a fast-forward path to `new`).

Since the slashing data is on a single `int` value on chain and might change for every new Tipset, we should walk the path and start registering when this value changes for each miner. Here I use a concept of `batchSize`, mostly to save calculated state every 20 Tipsets. If for a reason the process is killed, it only lost the some of the work done.